### PR TITLE
Fix bug 1377099 - Speed up status filters

### PR DESCRIPTION
**Pull request by adngdb**
There is one last test that is failing, and it's a mocking issue. I'll finish this next week, but that shouldn't have any impact on the actual code, so @mathjazz if you want to review this now, feel free! 

Note that I also haven't tested this with the Search across all projects feature yet.

### CODE REVIEW

File: pontoon/base/models.py [hunk 1]
```diff
@@ -1721,179 +1721,256 @@ def __str__(self):
...
+                self
+                .exclude(string_plural='')
+                .filter(pk__in=translations.values('entity'))
+                .prefetch_related(Prefetch(
+                    'translation_set',
+                    queryset=(
+                        Translation.objects
+                        .filter(locale=locale)
+                        .order_by('approved', 'fuzzy', '-date')
                     ),
```
Comment (mathjazz) [inline pontoon/base/models.py]: I think `I think `queryset=translations.order_by('approved', 'fuzzy', '-date')` would work here. And then you don't need `.filter(pk__in=translations.values('entity'))` above.

File: pontoon/base/models.py [hunk 2]
```diff
@@ -1721,179 +1721,256 @@ def __str__(self):
...
-                            translation__string=F('string_plural')
-                        ), then=1
-                    ), output_field=models.IntegerField(), default=0
+                    to_attr='fetched_translations'
+                ))
+            )
+
+            # Now that we have all those translations, we'll want to extract just the
+            # active one and then make sure it matches the `rule`. If it does, we store
+            # it to be retrieved in the finale query.
```
Comment (mathjazz) [inline pontoon/base/models.py]: Nit: final.

File: pontoon/base/models.py [hunk 3]
```diff
@@ -1721,179 +1721,256 @@ def __str__(self):
...
         )
 
-    def translated(self, locale, no_plurals):
-        if no_plurals:
-            approved = Translation.objects.filter(locale=locale, approved=True).values('entity')
-            return Q(pk__in=approved)
+    def rejected(self, locale):
+        """Return a filter to be used to select entities with rejected translations.
+
+        This filter will return any entities that have a rejected translation, whether
```
Comment (mathjazz) [inline pontoon/base/models.py]: s/any/all

**Comment by adngdb** [comment]
@mathjazz comments addressed, tested again with #520, this seems to be working great! And I've filed a bug for the unchanged filter issue.

File: pontoon/base/models.py [hunk 4]
```diff
@@ -1749,22 +1749,21 @@ def get_filtered_entities(self, locale, query, rule, match_all=True):
             # so that the first one for each plural form will be the active one.
             plural_candidates = (
                 self
+                .filter(
+                    resource__translatedresources__locale=locale,
+                    obsolete=False
+                )
```
Comment (mathjazz) [inline pontoon/base/models.py]: This makes things slow (e.g. slower than current master when filtering Firefox/All Resources).

If you revert 3f668c7 instead, we're good.

**Comment by adngdb** [comment]
@mathjazz reverting means a test fails, because of mocking issues. I spent a good part of the day fighting it, and I don't see any easy way of solving it without rewriting half the app, so I'm skipping it for now. :)

**Pull request merged**